bsi_zoo/cross_val.py

```diff
@@ -0,0 +1,194 @@
+#     negative log-likelihood.
+#     """
+#     log_lik = _gaussian_loglik_scorer(y, Sigma_Y)
+#     nll = -2 * log_lik
+#     return nll
+
+
+class BaseCVSolver(BaseEstimator, ClassifierMixin):
+    def __init__(
+        self,
+        solver,
+        cov_type,
+        cov,
+        n_orient,
+        alphas=np.linspace(1.4, 0.1, 20),
+        cv=5,
+        extra_params={},
+        n_jobs=1,
+    ):
+        self.solver = solver
+        self.alphas = alphas
+        self.cov = cov
+        self.cov_type = cov_type
+        self.n_orient = n_orient
+        self.cv = cv
+        self.extra_params = extra_params
+        self.n_jobs = n_jobs
+
+    def fit(self, L, y):
+        self.L_ = L
+        self.y_ = y
+
+        return self
+
+    def predict(self, y):
+        self._get_alpha(y)
+
+        if self.cov_type == "diag":
+            self.coef_ = self.solver(
+                self.L_,
+                y,
+                alpha=self.alpha_,
+                n_orient=self.n_orient,
+                **self.extra_params
+            )
+        else:
+            self.coef_ = self.solver(
+                self.L_,
+                y,
+                self.cov,
+                alpha=self.alpha_,
+                n_orient=self.n_orient,
+                **self.extra_params
+            )
+
+        return self.coef_
+
+
+class SpatialCVSolver(BaseCVSolver):
+    def _get_alpha(self, y):
+        """Sets alpha_ attribute with spatial cross-validation."""
+        gs = GridSearchCV(
+            estimator=SpatialSolver(
+                self.solver,
+                cov=self.cov,
+                alpha=None,
+                cov_type=self.cov_type,
+                n_orient=self.n_orient,
+            ),
+            param_grid=dict(alpha=self.alphas),
+            scoring="neg_mean_squared_error",
+            cv=self.cv,
+            n_jobs=self.n_jobs,
+        )
+        gs.fit(self.L_, y)
+        self.grid_search_ = gs
+        self.alpha_ = gs.best_estimator_.alpha
+
+
+class TemporalCVSolver(BaseCVSolver):
+    def _get_alpha(self, y):
+        """Sets alpha_ attribute with temporal cross-validation."""
+        base_solver = SpatialSolver(
+            self.solver,
+            cov=self.cov,
+            alpha=None,
+            cov_type=self.cov_type,
+            n_orient=self.n_orient,
+        )
+
+        cv = check_cv(self.cv)
+        scores = []
+        for alpha in self.alphas:
+            solver = clone(base_solver)
+            solver.set_params(alpha=alpha)
+            temporal_cv_scores = []
+            for train_idx, test_idx in cv.split(y.T):
+                solver.fit(self.L_, y[:, train_idx])
+                y_test = y[:, test_idx]
+                X_diag = np.sum(np.abs(solver.coef_), axis=1) != 0
+                # XXX this next formula seems to be wrong as it ignores
+                # the estimated or empirical source variance.
+                Sigma_Y = self.cov + (self.L_ * X_diag[None, :]) @ self.L_.T
```

Comment:
@AliHashemi-ai this seems very broken for now. You should need here the estimated gamma or compute an estimate of the source variance. What do you think?

Reply:
@agramfort  You are correct. I need to replace X_diag with the an estimate of the source variance. I will fix it.